### PR TITLE
ML-10 - Provide Activity Dates

### PR DESCRIPTION
**Pull request by dhansak79**
ML-10 - Provide Activity Dates

## Description

This pull request implements the complete test automation suite for **ML-10: Provide Activity Dates** feature, enabling applicants to provide start and end dates for their marine licensing activities. The implementation includes comprehensive BDD scenarios covering both happy path journeys and extensive validation testing.

### Purpose and Motivation:
- Enables users to specify when their marine activities will take place
- Implements robust date validation to ensure data quality and regulatory compliance
- Provides clear, specific error messaging for better user experience
- Supports the marine licensing application workflow progression

### Changes Introduced:
- **Complete BDD Test Suite**: Comprehensive Gherkin scenarios for activity dates functionality
- **Validation Framework**: Extensive validation testing for date inputs, including partial dates, invalid dates, past dates, and date order validation
- **Screenplay Pattern Implementation**: Full page objects, tasks, and interactions following established patterns
- **Enhanced Error Handling**: Specific error messages for individual date components (day, month, year)
- **User Story Documentation**: Detailed acceptance criteria and requirements documentation
- **Updates to Test Charters**: Reflect all user stories delivered so far

## Checklist

Before submitting the PR, ensure the following:

- [x] I have run all tests locally and confirmed they pass.
- [x] I have added tests that cover new functionality or changes (if applicable).
- [x] I have updated the documentation, including README.md and/or other relevant files.

## Related Tickets/Issues

- **Jira Ticket**: [ML-10](https://eaflood.atlassian.net/browse/ML-10) - Provide Activity Dates
- **Dependencies**: Builds upon ML-9 (View the task list)

## Changes

- [x] **Feature/Scenario(s) added**
  - `activity.dates.feature` - Happy path scenarios for providing activity dates
  - `validation.activity.dates.feature` - Comprehensive validation scenarios
- [x] **Step definitions added/modified**
  - Complete step definitions in `activity.dates.steps.js`
  - Enhanced error validation steps
- [x] **Core framework changes**
  - New page object: `activity.dates.page.js`
  - New model: `activity.dates.model.js`
  - New task: `complete.activity.dates.js`
  - Enhanced error handling capabilities

### 🎯 **Happy Path Scenarios**
- Successfully providing valid activity dates
- Pre-population of previously entered data
- Navigation to task list with completed status
- Back and cancel functionality

### ✅ **Comprehensive Validation Testing**
- **Mandatory Field Validation**: Empty start/end date handling
- **Partial Date Validation**: Specific errors for missing day, month, or year components
- **Invalid Date Validation**: Non-existent dates (e.g., 31st February, 30th April)
- **Past Date Validation**: Ensuring dates are today or in the future
- **Date Order Validation**: End date must be same as or after start date
- **Edge Cases**: Leap year validation, month boundary testing

### 🏗️ **Technical Implementation**
- **Page Object Pattern**: Clean separation of UI interactions
- **Screenplay Pattern**: High-level task abstraction for business logic
- **Model Layer**: Comprehensive date handling and validation logic
- **Error Handling**: Enhanced error message framework with specific validation types

### 📋 **Validation Error Messages**
The implementation provides specific, user-friendly error messages:
- `"Enter the start date"` / `"Enter the end date"` - for mandatory fields
- `"The start date must include a day"` - for missing day component
- `"The start date must include a month"` - for missing month component
- `"The start date must include a year"` - for missing year component
- `"The start date must be a real date"` - for invalid dates
- `"The start date must be today or in the future"` - for past dates
- `"The end date must be the same as or after the start date"` - for date order

### **Scenarios Covered**: 26 comprehensive test scenarios
- 6 happy path scenarios (main journey)
- 30 validation scenarios (comprehensive edge case coverage)

### **Key Review Areas**:
1. **Validation Logic**: Pay special attention to the comprehensive date validation scenarios
2. **Error Messages**: Verify alignment between feature files and user story requirements
3. **Framework Patterns**: Ensure consistency with established screenplay patterns
4. **Edge Cases**: Review the handling of leap years, month boundaries, and past dates

### **Notable Implementation Decisions**:
- **Specific Error Messages**: Moved from generic "must include day, month and year" to specific component-based messages
- **Error Message Stacking**: Support for multiple validation errors on the same field
- **Comprehensive Date Testing**: Extensive edge case coverage including leap years and month boundaries

## Screenshots (if applicable)

image

---

**This PR implements a complete, production-ready test suite for the Activity Dates feature, providing comprehensive coverage of both positive and negative test scenarios while maintaining high code quality and following established patterns.** 

[ML-10]: https://eaflood.atlassian.net/browse/ML-10?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test-infrastructure/screenplay/models/activity.dates.model.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
+import { faker } from '@faker-js/faker'
+
+export default class ActivityDatesModel {
+  static generateValidActivityDates() {
+    const startDate = faker.date.future({ years: 1 })
```
Comment (jonshi25) [inline test-infrastructure/screenplay/models/activity.dates.model.js:5]: 🏅 Great to faker generating data

File: test/features/validation.activity.dates.feature [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  Scenario Outline: Error when <condition> start date <startDay>, <startMonth>, <startYear> and <endDay>, <endMonth>, <endYear>
+    Given a notification has been created with a valid project name
+    And the activity start date has "<startDay>", "<startMonth>", "<startYear>"
+    And the activity end date has "<endDay>", "<endMonth>", "<endYear>"
+    When completing the activity dates task
+    Then the start date error "<expected_error>" is displayed
+
+    Examples:
+      | startDay | startMonth | startYear | endDay | endMonth | endYear | condition      | expected_error                                |
```
Comment (jonshi25) [inline test/features/validation.activity.dates.feature:34]: Why is expected_error in snake_case when everything else is in camelCase?

**Review by jonshi25 — APPROVED** [review]
LGTM 👍

File: test/features/validation.activity.dates.feature (hunk 2 above)
Comment (dhansak79) [reply test/features/validation.activity.dates.feature]: Sorted now - thanks!
